pennylane/devices/qubit/adjoint_jacobian.py

```diff
@@ -218,10 +218,20 @@ def adjoint_vjp(tape: QuantumTape, cotangents: Tuple[Number], state=None):
 
     ket = state if state is not None else get_final_state(tape)[0]
 
-    if np.shape(cotangents) == tuple():
-        cotangents = (cotangents,)
-    obs = qml.dot(cotangents, tape.observables)
-    bra = apply_operation(obs, ket)
+    cotangents = (cotangents,) if qml.math.shape(cotangents) == tuple() else cotangents
+    new_cotangents, new_observables = [], []
+    for c, o in zip(cotangents, tape.observables):
+        if not np.allclose(c, 0.0):
+            new_cotangents.append(c)
+            new_observables.append(o)
+    if len(new_cotangents) == 0:
+        return tuple(0.0 for _ in tape.trainable_params)
+    obs = qml.dot(new_cotangents, new_observables)
+    if obs._pauli_rep is not None:
+        flat_bra = obs._pauli_rep.dot(ket.flatten(), wire_order=list(range(tape.num_wires)))
```

Comment:
I am unsure whether this should be used whenever `obs._pauli_rep is not None`. This might do well in all cases, but it should do especially well for:
- Pauli words with several non-identity elements
- Pauli sentences with many such Pauli words
- Pauli sentences with Pauli words sharing the same sparsity structure (as is the case of QChem Hamiltonians)
I think I would leave it as is, but if performance decreases noticeably for certain workloads we can constrain use cases according to the above.